### DISCUSSION TITLE
Trouble with OIDC

### DISCUSSION THREAD

**Discussion by takuy** [directus/directus#10148] [Archive]
Hi. I'm having some trouble with my OIDC auth setup. I'm using https://github.com/simplesamlphp/simplesamlphp-module-oidc as an OIDC Connect provider as it helps bridge our existing SAML infrastructure. 

Running Directus 9.1.2 via Docker. 

```yaml
      AUTH_PROVIDERS: 'openid'
      AUTH_OPENID_DRIVER: 'openid'
      AUTH_OPENID_CLIENT_ID: 'snip'
      AUTH_OPENID_CLIENT_SECRET: 'snip'
      AUTH_OPENID_ISSUER_URL: 'https://mydomain.com/simplesaml/module.php/oidc/openid-configuration.php'
      AUTH_OPENID_SCOPE: 'openid profile email'
      AUTH_OPENID_IDENTIFIER_KEY: 'sub'
      AUTH_OPENID_ALLOW_PUBLIC_REGISTRATION: 'true'
```

The ISSUER URL contents are this:
![image](https://user-images.githubusercontent.com/1631095/143986615-3bbe5b65-c898-4362-96b6-41ad142b0c2f.png)

This configuration works in other scenarios - I'm just confused as to what's going wrong on Directus' end. When I use an OAuth config using the same provider, it kind of works? A user is inserted into the database, but I get an INVALID USER error. 

When I try it using OpenID - nothing happens, no insert, just an INVALID USER error. Here are the logs (`LOG_LEVEL = trace`) from the container:

```log
04:40:51 ✨ request completed GET 302 /auth/login/oidc?redirect=http://mydomain.com/directus/admin/login?continue 11ms
04:41:05 🔍 [1.114ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ⚠️  Invalid user credentials.
04:41:05 ✨ request completed GET 302 /auth/login/openid/callback?code=def502005ad1fceeeee5220196200fd6f0ad7803cae6467637748936e41f58ec4cb5db46deec53e7feb833a10b109eae50a9b5c6f7464180b3497975c35cba02e0a9ac32639f7ed7f95d0202aba04415c4b38472174d0b1a2369197b5dcdc8a0348bc0bcf15bc1837dae0276818d83313cc86382ef5c808c66583b96f2372d581a5f71114be12cf03255f26c383fce3f1644bb0e6b51a4ef497acf4983dedbf394e7d860403beade14339f709d7f817af83b42b3d5b8ec47aa4d8547a3cf88b43081bdf688e5a7be9f8b51e4ae40d207dcb763eb49b68fe80c08c253004afb5fb59a01a1b83e4d9793d66526ce3680372691a6e03faf9634db76962289e25ad05b6adf0ae8cfcbdd7a8508d18e5b913fd98e30032005ceae98bb878c500533a87e8f12f5c49ed84e0fc764eb6a3c4052ccf0ceea763afaa407b6966dba2daeaaa8b69151460f3fd0b5e2566ed9e7bfba17121ca9204a36529bdc5513df0712b11dcff60f8c5dcec2a91eb6d611589f773ed989fc7f7910bfea41755dd21bb8421436112f795b6256a0684b00852eea23cc15e0c6dc4927a73529944ec707f2143d66489352aa578e249513d8d95fde5e3bd0e227013b2472f48d75d534d65645a3c43cc89975353bdd2583d46d67c7f4bed598950c14644dae36fcef4860acb58427d601b4817e670220a4a67c2a58afbee99aa403421b30decce30ac28292652fc2d1354481f800b9d5bdf6ad5ccd3f574c33feb7f38339d61e3e278af455247f8a30c9de7350a305ee027a77ede3afa5cc8035eb&state=dUJeny5zhmcGJOOhRkLBFc5bKTBRYgnw2KND83p_3b8 418ms
04:41:05 ✨ request completed GET 304 /admin/login?reason=INVALID_USER 1ms
04:41:05 ✨ request completed GET 304 /admin/index.b5d16ba8.js
04:41:05 ✨ request completed GET 304 /admin/assets/vendor.fe7856f0.css 2ms
04:41:05 ✨ request completed GET 304 /admin/assets/index.972d7652.css 1ms
04:41:05 ✨ request completed GET 304 /admin/assets/vendor.c46e17d6.js 2ms
04:41:05 ✨ request completed GET 304 /admin/assets/use-sync.864aa3cf.js 1ms
04:41:05 ✨ request completed GET 304 /admin/assets/vendor.c46e17d6.js 2ms
04:41:05 ✨ request completed GET 304 /admin/assets/use-sync.864aa3cf.js 2ms
04:41:05 🔍 [3.263ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ✨ request completed GET 200 /extensions/interfaces/index.js 10ms
04:41:05 🔍 [8.556ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ✨ request completed GET 200 /extensions/panels/index.js 12ms
04:41:05 🔍 [1.878ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ✨ request completed GET 200 /extensions/displays/index.js 6ms
04:41:05 🔍 [1.792ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ✨ request completed GET 200 /extensions/modules/index.js 8ms
04:41:05 🔍 [12.703ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 ✨ request completed GET 200 /extensions/layouts/index.js 17ms
04:41:05 🔍 [5.986ms] select `ip_access` from `directus_roles` where `id` is null limit ? [1]
04:41:05 🐛 "refresh_token" is required in either the JSON payload or Cookie
04:41:05 ✨ request completed POST 400 /auth/refresh 17ms
```

Where can I start with debugging this?

**Comment by aidenfoxx** [comment]
I get the feeling the problem is that the issuer in your `id_token` response doesn't match the issuer in your `ISSUER_URL` config. The library we're using requires the `id_token` to have the correct `iss` value.

This recently occurred for another user who was implementing a custom OpenID provider: https://github.com/directus/directus/discussions/9066#discussioncomment-1683768

**Reply by takuy** [reply]
Is there a way to get verbose output for the OAuth flow & responses on Directus end? 

I'll investigate on the OIDC provider end to see what's going on as well but seeing exactly what Directus gets as a response would be helpful. Was surprised that trace level logging didn't help at all.

**Reply by aidenfoxx** [reply]
Yeah, there aren't many trace logs in Directus. We should probably add more. 🤷 

It also doesn't look like `openid-client` does much trace logging either.

**Reply by takuy** [reply]
So, I was able to add some debug output on the provider end to grab the response from the authorize token endpoint. The issuer in the well-known config and the ISS in the id_token are exactly the same. 

The access_token in the authorization response even works fine against the userinfo endpoint, but honestly Directus doesn't even seem to be getting that far. 

The payload looks like this:

```json
{
  "iss": "https://mydomain.com",
  "iat": 1638297699,
  "jti": "fe8b4bb5f7a660ba4485ccf7aa64dbe9cdc19df56de07f34a980cc048da557b933fc1b8aa028a2ad",
  "aud": "my client id",
  "nbf": 1638297699,
  "exp": 1638301299,
  "sub": "myorgusername",
  "at_hash": "Ohfz6e_80XWrcPUZZBdx8g",
  "sid": "8c915ec008679cd1ddade22719094f1f"
}
```

Am I missing anything?

**Reply by aidenfoxx** [reply]
Would be nice to know what's happening in the `handleError` function. If you could manually add a log to the `node_modules/directus/dist/auth/drivers/openid.js` file it might help.

**Reply by takuy** [reply]
`19:51:29 🔍 error: RPError: failed to validate JWT signature`

🤔

**Reply by aidenfoxx** [reply]
Not sure what to tell you. Something in your JWT is failing validation in `openid-client`: https://github.com/panva/node-openid-client/blob/b580dc53552ee080ed55af6c6897f54ff786d897/lib/client.js#L1010

**Reply by takuy** [reply]
Question, forgive my ignorance on the subject, does that function validate the id_token, the access_token, and then refresh_token returned by the authorizeToken endpoint and return an error if any of them are invalid?

The refresh token I'm getting returned *is not* a jwt, which is a bit weird.

**Reply by aidenfoxx** [reply]
I just think it just validates the id_token which should be a jwt, but I can't check the code right now.

**Reply by takuy** [reply]
Okay, I've gotten further. There was an issue on the provider side - the key wasn't the correct length for whatever reason. That's resolved. The OIDC process is getting further. It's doing a weird thing though. 

It does the whole authorizeToken process, writes the access token and refresh token to the database, creates/updates the user data, looks all good in the logs... and then it uses the refresh_token in a refresh flow, gets a new token, which causes the old one to invalidate, and then tries to do another refresh flow with the old refresh_token? 

The only error from Directus is that the 2nd time it tries to do a refresh token, the token is invalid. 

This is what I'm seeing in the provider logs:
```log
Dec 01 01:07:51 simplesamlphp DEBUG [21cd28ea1b] Token Request: {"grant_type":"authorization_code","code":"<snip>","redirect_uri":"http:\/\/mydomain.com\/directus\/auth\/login\/sso\/callback","code_verifier":"<snip>"}
Dec 01 01:07:51 simplesamlphp DEBUG [21cd28ea1b] Token Response: {"id_token":"<snip>","token_type":"Bearer","expires_in":3600,"access_token":"<snip>","refresh_token":"<refreshtoken1>"}
Dec 01 01:07:51 simplesamlphp DEBUG [61a0e80b4b] Token Request: {"grant_type":"refresh_token","refresh_token":"<refreshtoken1>"}
Dec 01 01:07:51 simplesamlphp DEBUG [61a0e80b4b] Token Response: {"id_token":"<new id token>","token_type":"Bearer","expires_in":3600,"access_token":"<new access token>","refresh_token":"<refreshtoken2>"}
Dec 01 01:07:52 simplesamlphp DEBUG [105214b31c] Token Request: {"grant_type":"refresh_token","refresh_token":"<refreshtoken1>"}
Dec 01 01:07:52 simplesamlphp ERROR [105214b31c] SimpleSAML\Module\oidc\Server\Exceptions\OidcServerException: The refresh token is invalid.
```

So, I'm still pretty confused. Everything looks like it should be working.

**Reply by aidenfoxx** [reply]
So the problem is that your app is returning a "refresh_token" on the refresh grant call. We just expect an "access_token" to be returned.

The oauth specs are loose at the best of times, and I've never seen an example where a refresh grant returns a new "refresh_token", but it is something we should handle and store the new "refresh_token".

Although your refresh grant doesn't return an "access_token", which seems wrong.

**Reply by takuy** [reply]
The refresh grant does return an access_token:
` Dec 01 01:07:51 simplesamlphp DEBUG [61a0e80b4b] Token Response: {"id_token":"<new id token>","token_type":"Bearer","expires_in":3600,"access_token":"<new access token>","refresh_token":"<refreshtoken2>"}`

**Comment by aidenfoxx** [comment]
So yes, according to https://datatracker.ietf.org/doc/html/rfc6749#section-6 we should be updating the refresh token if one is provided:
```
The authorization server MAY issue a new refresh token, in which case the client MUST discard the old refresh token
```

**Reply by takuy** [reply]
Should I submit an issue for this?

**Reply by aidenfoxx** [reply]
I plan on fixing it over the weekend, along with implement other improvements. You can create an issue if you wish to track it.
